### ISSUE TITLE
[Discovery]  Veteran is registered at one or more facilities, and more than one facility has EC enabled and available today

### ISSUE THREAD

**Issue by lenaecb** [department-of-veterans-affairs/va.gov-team#11412]
**Use case:** Veteran is registered at one or more facilities, and more than one facility has EC enabled and available today

**Example:** Veteran is registered at San Fransisco and Sacramento and both have Express Care enabled for today. 

**Considerations:**
- We can ask the Veteran to select from sites that are enabled/available today

## Acceptance criteria:
- Walk through desired user flow for this scenario
- Design mocks for handling this flow throughout EC flow - from homepage to submission

**Comment by rtwell** [comment]
i wonder if it is its own screen like this (these words are terrible):

![image](https://user-images.githubusercontent.com/12564977/88311190-94d4c700-ccde-11ea-8924-2839fddf6229.png)

**Comment by lenaecb** [comment]
Per conversation with Lauren - this will not be part of our v1 rollout. This _is_ a use case to account for and decide how we want to solve for it. So, ideation is still needed.

**Comment by lenaecb** [comment]
@lalexanderson-dsva @rtwell - this will need to be brought up to stakeholder next week. Current iteration of EC in VAOSR (v1) _does not include this use case._

What we need to confirm with stakeholders is if:
- we can release v1 of EC in VAOSR _without this use case_
- when we need to account for this use case (Wilmington is estimated to come online in the next few weeks so worth solving is use case in coming sprint - sprint 23)

**Comment by lalexanderson-dsva** [comment]
@lenaecb  - if they do not have it now, why do we need to account for it in MVP? 

This is really making me want the "set my preferred facility" option we were discussing in our 11am meeting this morning.

**Comment by jbalboni** [comment]
@lalexanderson-dsva They do have it in legacy; you have to choose your city/state and facility, if there are multiple options.

In the v1 on va.gov, the way the first pass is coded is that we'll pick whatever facility with an open window that we happen to see first. If only SF and Wilmington are turned on, you'd have to be registered at both sites and be checking at a time both are accepting EC requests. That's probably not a common use case.

**Comment by lenaecb** [comment]
**Decision based on Design/Content sync:** If Veteran has multiple active windows for EC at one time, we will default select facility closest in proximity to Veterans home address who has EC enabled. Long term, Veteran will have the option to designate primary/preferred facility in their VA Profile, but that is not yet on the VA Profile team's roadmap. 

**So, Veteran who is eligible for EC at multiple sites:**
- VAOS will check for closest facility to Veteran's residential/home address from VA Profile and choose facility closest to that address
- VAOS will automatically also check request limit for EC and that facility as part of the facility selection
- Prompt that displays to Veteran (whether active or inactive) will display based on closest facility to Veteran location and display that window/time for EC if within request limit
   - if outside of request limit for facility with EC, check next available facility with EC enabled and display that time/window

**Need to create following issues:**
1. BE service update request - for facilities and request limit automated checks
2. FE - update to select active EC facility closest to Veteran's address

CC: @lalexanderson-dsva @rtwell - add anything additional you want considered here.

**Comment by lenaecb** [comment]
cc: @narin @jbalboni see comment above with our desired approach. 

Thoughts? Feedback?

Also, have we already submitted a ticket to update BE with the request limit check?

**Comment by jbalboni** [comment]
I _think_ that is feasible, but it's definitely going to delay supporting multi-site users. My concerns are:

1. I'm not sure we know if the cost of those service calls to get distances is feasible or who will pay for them. We might be in the Mapbox free tier initially, but if EC becomes broadly used, we'd easily be in the hundreds of dollars a month range, if not more. We also have to account for rate limits and make sure we're under those.
2. There's a possibility this will be too slow for use on the homepage. We're looking at at least 4 service calls, 3 of which will have to be done in sequence. We have to:

1. Fetch the currently supported EC sites (VAMF)
2. In parallel:
  - Get coordinates of current address (Mapbox, because I don't believe this is in VA Profile)
  - Get coordinates of supported facilities (VA facilities api)
  - Check request limits for supported facilities (VAMF)
3. Call Mapbox Matrix API to get distances between points

The way things currently work, that would block the loading of the list tabs while in progress, and follows on the already pretty slow setup calls for authentication and feature toggles we have to wait for.

I do believe there's a ticket for updating the request limit call.

Additionally, while we can do this in the FE, it's really work that belongs in the VAOS backend, based on the previously stated desires of the team to have business logic live in the VAMF backend.

**Comment by rtwell** [comment]
Thanks for the extensive feedback, Jeff!

It would be helpful (at least, to me) if you and Narin could give us your collective engineering recommendation for the most practical, efficient way for the backend to handle this facility/time/request problem, review that together and make the recommendation to the VAMF team. 

We opted for the described approach above to avoid giving some veterans more access to care than others, and (what I thought was) a straightforward, rigid path of only ever having access to 1 facility (or, 1 instance of EC). How could we pull that off without incurring the delays you described? I thought we already ran a mapbox call for drivetime anyway, so i assumed we could get away with another (dammit). 

i'm also trying to avoid forcing the Veteran to parse facilities, time zones, and whatnot—if we could somehow only surface a single EC instance, that would be ideal.

**Comment by jbalboni** [comment]
Another option would be to use the VA facilities api' s nearby call and pull facilities within a certain distance. You could use the intersection between that list and the list of facilities supporting EC to find the nearest ones. It wouldn't be perfect, but it might be good enough.

We could also check with the lighthouse team about extending the nearby call to accept a list of facilities, or a new endpoint with similar behavior, which would cover all our needs.

**Comment by jbalboni** [comment]
Like I said, I think we have feasible options here, we just have to do some discovery and make sure the effort and impacts are acceptable to everyone.

**Comment by jbalboni** [comment]
Hi @davem-va, I'm curious if you have thoughts about if the VA facilities api could support the scenario I was thinking about above: 

Essentially, we'd like to be able to send an address and a list of facility ids, and get back the drive time distances for each facility from the address we sent. Pretty much the same functionality as `/nearby` but with a facility id list instead of a distance.

**Comment by davem-va** [comment]
Hey @jbalboni   Would it be possible for you to send us geocodes instead of an address?  We currently don't use any mapping service and would need one to provide drive times.

**Comment by jbalboni** [comment]
@davem-va Possibly, but doesn't `nearby` take an address now? What would be different here?

**Comment by davem-va** [comment]
It does, but this is something we were on the way to removing.  PSSG is the authoritative source of geocoding for health care enrollees.  The CCE API uses the Enrollment and Eligibility API to get this geocoding from PSSG and uses it in calls to /nearby.

What problem are we trying to solve in passing the address rather than geocodes to /nearby?

**Comment by jbalboni** [comment]
@davem-va We're trying to avoid a mapping service, same as you all. 😄 

There's some conflicting info on the Mapbox site, but I think we could use their api like the facility locator does to get geocodes to send.

That said, we're not trying to do this for any arbitrary address, it would be the address of the current user on VAOS. Would it be feasible to have a service where we sent an ICN and a list of facilities, and got back drive time bands for each, based on the enrollee geocodes you get from PSSG?

**Comment by davem-va** [comment]
Thanks Jeff.  I guess the larger question is should anyone other than PSSG (and the underlying ESRI mapping service they use) supply geocodes for a given address belonging to a health care enrollee?    I believe this answer to be no.

**Comment by jbalboni** [comment]
@davem-va Yeah, I think that makes sense. So that leaves us with deciding if we (VAOS) need to get them directly from PSSG, or if that step could be incorporated into the theoretical new lighthouse API I'm asking about.

**Comment by davem-va** [comment]
In either approach, I think we need an API from PSSG for the translation if the geocoding they provide via "normal means" isn't available through E&E.  We've asked for one in the past.  They don't have one and didn't have the staff/budget to build one.

Such an API could also solve another issue as well, namely the delay between an address update coming from va-profile into E&E and PSSG doing their 2x daily geocoding for new/updated addresses.  The VA-profile team was working this and the fact that geocoding for non-health-care enrollees is done with a different mapping service.  I'm not sure where it stands.

**Comment by jbalboni** [comment]
Why would that API be required here, if it's not required for doing the drive time check for CC?

Might be worth syncing up so I can understand this better and stop asking you endless questions in here. 😄

**Comment by lenaecb** [comment]
Capturing some follow up and next steps here from sync with Dave M. 
- Dave intro us to Nancy about PSSG data & extending drive time beyond 90mins (current cap)
- VAOSR team - validate geocodes and if they are exposed with VA Profile team
  - see if we can get documentation on what's available
  - ask if they cane expose the codes if not already exposed
- what is an max distance for veterans to facilities

**Comment by lenaecb** [comment]
@jbalboni - based on what we're looking to do with the location selection (referencing facilities based on residential address), I would assume we'd repurpose that logic for what we're trying to do here for multi-site EC windows?

**Comment by jbalboni** [comment]
@lenaecb Probably? I'd actually be interested to think through if we want slightly different logic here because EC is generally a phone/video appointment and because we're choosing a single facility. But we could definitely use the same logic here as on the facilities page if we want.

**Comment by lenaecb** [comment]
@jbalboni - sorry, I just meant in terms of sorting/selecting the active EC facility (when there is more than 1) by location proximity. Does the other work we're considering for the location selection page not also help us do that type of sorting? What else did you have in mind?

**Comment by jbalboni** [comment]
@lenaecb I think we're on the same page, I just wasn't being very clear in my response. We have the technical ability to do several different things:

- Sort facilities by straight line distance from residential address
- Sort facilities within 90 minutes by drive time from residential address
- Use current location from browser (after asking permission) as the starting point for sorting distances/drive times

How we use those capabilities in the VA facility choice context and in the EC context is totally up to you all. I didn't want to assume that we'd use the exact same logic in both contexts, since there are differences in what we're trying to accomplish, and I'm not a product person. I don't have anything specific in mind that we should do differently between the two contexts.

**Comment by lenaecb** [comment]
Ah gotcha, yes same page. Thanks for adding those tech options, super helpful @jbalboni 

I think we go with both highest value and lightest lift for EC because of the limited usage of the feature now. There are still a lot of open questions such as - ease of data transfer between facilities and how many EC requests actually turn into f2f appts. We won't know that for awhile because so few sites are using this + small number of EC requests.

That being said, my recommendation would be:
- Sort facilities by straight line distance from residential address
- because this ensures we properly account for outliers (folks in rural areas) that are beyond 90mins

@rtwell @lalexanderson-dsva - do you agree with my recommendation? This is for there are multiple EC windows open at the same time.

**Comment by lenaecb** [comment]
Based on design/content sync - move forward with this option - "Sort facilities by straight line distance from residential address"

We will revisit open questions (relating more to scalability of EC) in the future when/if feature adoption continues.
